v2/services/authorize/client_retry_test.go

```diff
@@ -0,0 +1,345 @@
+
+func parseRSAKey() (*rsa.PrivateKey, error) {
+	pemBlock, _ := pem.Decode(rsaKey)
+	k, err := x509.ParsePKCS8PrivateKey(pemBlock.Bytes)
+	if err != nil {
+		return nil, err
+	}
+	return k.(*rsa.PrivateKey), nil
+}
+
+type server struct {
+	signal chan int
+}
+
+func launchServer(tlsCredentials credentials.TransportCredentials) (server, error) {
+	server := server{
+		signal: make(chan int),
+	}
+
+	var serverOpts []grpc.ServerOption
+	serverOpts = append(serverOpts, grpc.Creds(tlsCredentials))
+
+	grpcServer := grpc.NewServer(serverOpts...)
+	authorizeproto.RegisterAuthorizeServer(grpcServer, newServer())
+
+	lis, err := net.Listen("tcp", "localhost:10000")
+	if err != nil {
+		return server, err
+	}
+
+	go func() {
+		//nolint:errcheck
+		go grpcServer.Serve(lis)
+
+		<-server.signal
+
+		grpcServer.Stop()
+		_ = lis.Close()
+		server.signal <- 1
+	}()
+
+	return server, nil
+}
+
+func (s *server) Shutdown() {
+	s.signal <- 1
+	<-s.signal
+	close(s.signal)
+}
+
+func TestDefaultDeadline(t *testing.T) {
+	privateKey, err := parseRSAKey()
+	require.NoError(t, err)
+
+	caCertPEM, err := generateCA(privateKey)
+	require.NoError(t, err)
+
+	serverDataStore, err := generateDatastore(ca, privateKey, caCertPEM, 10*365*24*time.Hour)
+	require.NoError(t, err)
+
+	clientDataStore, err := generateDatastore(ca, privateKey, caCertPEM, 3*24*time.Hour)
+	require.NoError(t, err)
+
+	tlsCredentials, err := loadTLSCredentials(serverDataStore)
+	require.NoError(t, err)
+
+	server, err := launchServer(tlsCredentials)
+	require.NoError(t, err)
+
+	c := authorize.CreateClient()
+	c.SetRequestTimeout(time.Millisecond)
+
+	err = c.DialUsingCredentialsManager(context.Background(), &mockCredentialsFetcher{ds: clientDataStore}, "localhost", "10000", "")
+	require.NoError(t, err)
+
+	server.Shutdown()
+
+	_, err = c.GetResourceWithContext(context.Background(), "", "")
+
+	require.EqualError(t, err, "rpc error: code = DeadlineExceeded desc = context deadline exceeded")
+}
+
+func TestReconnect(t *testing.T) {
+	privateKey, err := parseRSAKey()
+	require.NoError(t, err)
+
+	caCertPEM, err := generateCA(privateKey)
+	require.NoError(t, err)
+
+	serverDataStore, err := generateDatastore(ca, privateKey, caCertPEM, 10*365*24*time.Hour)
+	require.NoError(t, err)
+
+	clientDataStore, err := generateDatastore(ca, privateKey, caCertPEM, 24*time.Hour)
+	require.NoError(t, err)
+
+	tlsCredentials, err := loadTLSCredentials(serverDataStore)
+	require.NoError(t, err)
+
+	server, err := launchServer(tlsCredentials)
+	require.NoError(t, err)
+
+	c := authorize.CreateClient()
+
+	err = c.DialUsingCredentialsManager(context.Background(), &mockCredentialsFetcher{ds: clientDataStore}, "localhost", "10000", "")
+	require.NoError(t, err)
+
+	server.Shutdown()
+
+	server, err = launchServer(tlsCredentials)
+	require.NoError(t, err)
+	defer server.Shutdown()
+
+	_, err = c.GetResourceWithContext(context.Background(), "", "")
+
+	require.NoError(t, err)
+}
+
+func TestRetryPolicy(t *testing.T) {
```

Comment:
Does this test really test the policy?

Reply:
Yes is does. Updated to make it more clear.